### DISCUSSION TITLE
How to add a custom compiler path at runtime?

### DISCUSSION THREAD

**Discussion by ClockworkV** [compiler-explorer/compiler-explorer#4702] [General]
I'm working on a team developing a custom compiler. 
We would love to have a way of comparing output of some baseline version and the currently developed version. 
I could hardcode it into some properties file, but then we'd get a huge list of compilers.
Ideally I  could select the compiler as "Custom version" and provide a path to the executable.
What would be the path of least resistance to adding such a feature?

I started trying to add a "compilerPath" field next to the "Option" field in the compiler view top bar, and trying to change the .exe path. This didn't immediately work, and I'm wondering whether to pursue this direction, or try something else. 
Another option could be to define new compilers, via the web interface, per user.

Has anyone done something similar, or has a clear idea of how to do this? I'm somewhat familiar with Typescript, so I'm not looking for someone else to actually do the work, just directions.

Cheers!

Vladi

**Comment by mattgodbolt** [comment]
We haven't ever considered this use case before -- to do this on our side we point "trunk" builds at a symlink, and then switch the symlink. Does that work for you as a start point?

**Reply by ClockworkV** [reply]
This would work I think, I could have a symlink per developer. Not the best ergonomics, but you need to start somewhere (: 
Does this work well? With regard to caching results, displaying the compiler path etc? Say, when I rebuild the compiler, or point the symlink at another one?

**Reply by AbrilRBS** [reply]
Hi! Sorry for the radio silence, we are a it busy lately and some issues fall thru the cracks :/

Regarding your question, I don't know if you ended up trying, but I wouldn't expect the symlink to work nicely with caching by default, you'd need to at least make sure you press the Refresh button in the compiler pane to force flush the cache. As for the shown path, the symlink will not be resolved for the UI, no. Did you end up using symlinks? How did they work? Do you feel that going with your original idea could be a better approach for you and your team? If so, let us know and I'll be happy to help implementing it where possible :)

**Reply by ClockworkV** [reply]
Hi! Thanks for even responding, it's not at all taken for granted!

At the moment, we're using the symlink approach. The one colleague who's using it the most says he has to reload the website when changing the symlink for it to work. But it's no big deal. 
Having thought about it, it seems that the most common scenario is that a person needs to define a new 'compiler' once and then use that a lot. So there wouldn't be a lot of "change the path to the compiler on the fly" type of thing. In terms of user interaction, perhaps the best would be something that allows the user to define their own compilers, and have them added to the list. Something like a per-user defaults.properties file. 
I haven't done a lot of work in this direction, as this is a side project I'm working in a background thread (: I'll try to push upstream if anything useful comes out of it.
